### PR TITLE
Add branch detection to the semantic model

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

We have a few rules that rely on detecting whether two statements are in different branches -- for example, different arms of an `if`-`else`. Historically, the way this was implemented is that, given two statement IDs, we'd find the common parent (by traversing upwards via our `Statements` abstraction); then identify branches "manually" by matching the parents against `try`, `if`, and `match`, and returning iterators over the arms; then check if there's an arm for which one of the statements is a child, and the other is not.

This has a few drawbacks:

1. First, the code is generally a bit hard to follow (Konsti mentioned this too when working on the `ElifElseClause` refactor). 

2. Second, this is the only place in the codebase where we need to go from `&Stmt` to `StatementID` -- _everywhere_ else, we only need to go in the _other_ direction. Supporting these lookups means we need to maintain a mapping from `&Stmt` to `StatementID` that includes every `&Stmt` in the program. (We _also_ end up maintaining a `depth` level for every statement.) I'd like to get rid of these requirements to improve efficiency, reduce complexity, and enable us to treat AST modes more generically in the future. (When I looked at adding the `&Expr` to our existing statement-tracking infrastructure, maintaining a hash map with all the statements noticeably hurt performance.)

The solution implemented here instead makes branches a first-class concept in the semantic model. Like with `Statements`, we now have a `Branches` abstraction, where each branch points to its optional parent. When we store statements, we store the `BranchID` alongside each statement. When we need to detect whether two statements are in the same branch, we just realize each statement's branch path and compare the two. (Assuming that the two statements are in the same scope, then they're on the same branch IFF one branch path is a subset of the other, starting from the top.) We then add some calls to the visitor to push and pop branches in the appropriate places, for `if`, `try`, and `match` statements.

Note that a branch is not 1:1 with a statement; instead, each branch is closer to a suite, but not _every_ suite is a branch. For example, each arm in an `if`-`elif`-`else` is a branch, but the `else` in a `for` loop is not considered a branch.

In addition to being much simpler, this should also be more efficient, since we've shed the entire `&Stmt` hash map, plus the `depth` that we track on `StatementWithParent` in favor of a single `Option<BranchID>` on `StatementWithParent` plus a single vector for all branches. The lookups should be faster too, since instead of doing a bunch of jumps around with the hash map + repeated recursive calls to find the common parents, we instead just do a few simple lookups in the `Branches` vector to realize and compare the branch paths.

## Test Plan

`cargo test` -- we have a lot of coverage for this, which we inherited from PyFlakes

### CODE REVIEW

File: crates/ruff_python_semantic/src/analyze/branch_detection.rs [hunk 1]
```diff
@@ -1,105 +0,0 @@
...
-        for items in alternatives(node_tree[ancestor]) {
-            let l = descendant_of(left, &items, ancestor, node_tree);
-            let r = descendant_of(right, &items, ancestor, node_tree);
-            if l ^ r {
-                return true;
-            }
-        }
-    }
-    false
-}
```
Comment (charliermarsh) [inline crates/ruff_python_semantic/src/analyze/branch_detection.rs:105]: Glad to see this removed.

File: crates/ruff_python_semantic/src/statements.rs [hunk 2]
```diff
@@ -22,70 +21,41 @@ struct StatementWithParent<'a> {
...
-    depth: u32,
+    /// The branch ID of this node, if any.
+    branch: Option<BranchId>,
 }
 
 /// The statements of a program indexed by [`StatementId`]
 #[derive(Debug, Default)]
 pub struct Statements<'a> {
     statements: IndexVec<StatementId, StatementWithParent<'a>>,
-    statement_to_id: FxHashMap<StatementKey, StatementId>,
```
Comment (charliermarsh) [inline crates/ruff_python_semantic/src/statements.rs:33]: Glad to see this go.

**Comment by charliermarsh** [comment]
Small improvement on benchmarks (no decreases in performance, `large/dataset.py` improved by 1-2%):

```
linter/default-rules/numpy/globals.py
                        time:   [60.583 µs 60.774 µs 61.010 µs]
                        thrpt:  [48.363 MiB/s 48.552 MiB/s 48.705 MiB/s]
                 change:
                        time:   [-0.3999% -0.0452% +0.3162%] (p = 0.81 > 0.05)
                        thrpt:  [-0.3152% +0.0452% +0.4015%]
                        No change in performance detected.
Found 15 outliers among 100 measurements (15.00%)
  6 (6.00%) high mild
  9 (9.00%) high severe
Benchmarking linter/default-rules/pydantic/types.py: Collecting 100 samples in estimated 25.061 s (20k ite
linter/default-rules/pydantic/types.py
                        time:   [1.2391 ms 1.2398 ms 1.2406 ms]
                        thrpt:  [20.557 MiB/s 20.571 MiB/s 20.583 MiB/s]
                 change:
                        time:   [-1.3773% -1.1253% -0.8898%] (p = 0.00 < 0.05)
                        thrpt:  [+0.8978% +1.1381% +1.3966%]
                        Change within noise threshold.
Found 8 outliers among 100 measurements (8.00%)
  5 (5.00%) high mild
  3 (3.00%) high severe
Benchmarking linter/default-rules/numpy/ctypeslib.py: Collecting 100 samples in estimated 20.407 s (35k it
linter/default-rules/numpy/ctypeslib.py
                        time:   [578.64 µs 580.74 µs 583.33 µs]
                        thrpt:  [28.545 MiB/s 28.673 MiB/s 28.776 MiB/s]
                 change:
                        time:   [-1.0377% -0.7648% -0.4813%] (p = 0.00 < 0.05)
                        thrpt:  [+0.4836% +0.7707% +1.0486%]
                        Change within noise threshold.
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe
Benchmarking linter/default-rules/large/dataset.py: Collecting 100 samples in estimated 59.321 s (20k iter
linter/default-rules/large/dataset.py
                        time:   [2.9103 ms 2.9139 ms 2.9192 ms]
                        thrpt:  [13.936 MiB/s 13.962 MiB/s 13.979 MiB/s]
                 change:
                        time:   [-2.1756% -2.0463% -1.9141%] (p = 0.00 < 0.05)
                        thrpt:  [+1.9514% +2.0890% +2.2239%]
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  6 (6.00%) high mild
  4 (4.00%) high severe

Benchmarking linter/all-rules/numpy/globals.py: Collecting 100 samples in estimated 10.536 s (86k iteratio
linter/all-rules/numpy/globals.py
                        time:   [121.56 µs 121.70 µs 121.84 µs]
                        thrpt:  [24.217 MiB/s 24.246 MiB/s 24.273 MiB/s]
                 change:
                        time:   [-1.5244% -1.0506% -0.7164%] (p = 0.00 < 0.05)
                        thrpt:  [+0.7216% +1.0618% +1.5480%]
                        Change within noise threshold.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe
Benchmarking linter/all-rules/pydantic/types.py: Collecting 100 samples in estimated 25.274 s (10k iterati
linter/all-rules/pydantic/types.py
                        time:   [2.4946 ms 2.4955 ms 2.4965 ms]
                        thrpt:  [10.216 MiB/s 10.220 MiB/s 10.223 MiB/s]
                 change:
                        time:   [-1.2063% -0.9096% -0.6549%] (p = 0.00 < 0.05)
                        thrpt:  [+0.6592% +0.9179% +1.2210%]
                        Change within noise threshold.
Found 13 outliers among 100 measurements (13.00%)
  8 (8.00%) high mild
  5 (5.00%) high severe
Benchmarking linter/all-rules/numpy/ctypeslib.py: Collecting 100 samples in estimated 25.683 s (20k iterat
linter/all-rules/numpy/ctypeslib.py
                        time:   [1.2649 ms 1.2658 ms 1.2670 ms]
                        thrpt:  [13.143 MiB/s 13.155 MiB/s 13.164 MiB/s]
                 change:
                        time:   [+0.1974% +0.7212% +1.3268%] (p = 0.01 < 0.05)
                        thrpt:  [-1.3094% -0.7160% -0.1970%]
                        Change within noise threshold.
Found 13 outliers among 100 measurements (13.00%)
  4 (4.00%) high mild
  9 (9.00%) high severe
Benchmarking linter/all-rules/large/dataset.py: Collecting 100 samples in estimated 50.342 s (10k iteratio
linter/all-rules/large/dataset.py
                        time:   [4.9667 ms 4.9736 ms 4.9828 ms]
                        thrpt:  [8.1646 MiB/s 8.1797 MiB/s 8.1912 MiB/s]
                 change:
                        time:   [-1.7667% -1.5446% -1.3479%] (p = 0.00 < 0.05)
                        thrpt:  [+1.3663% +1.5688% +1.7985%]
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  5 (5.00%) high mild
  6 (6.00%) high severe
```

**Comment by charliermarsh** [comment]
(I don't think any of the benchmarks even trigger the branch detection though, that's just from not hashing the statement etc. presumedly.)

**Comment by charliermarsh** [comment]
I replaced one of the benchmarks with a file that includes a lot of redefinitions-within-branches, small improvement:

```
linter/default-rules/numpy/globals.py
                        time:   [60.227 µs 60.337 µs 60.457 µs]
                        thrpt:  [48.806 MiB/s 48.903 MiB/s 48.993 MiB/s]
                 change:
                        time:   [-1.6113% -1.3619% -1.1125%] (p = 0.00 < 0.05)
                        thrpt:  [+1.1251% +1.3807% +1.6377%]
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

Benchmarking linter/all-rules/numpy/globals.py: Collecting 100 samples in estimated 10.464 s (86k iteratio
linter/all-rules/numpy/globals.py
                        time:   [121.33 µs 121.46 µs 121.60 µs]
                        thrpt:  [24.266 MiB/s 24.294 MiB/s 24.320 MiB/s]
                 change:
                        time:   [-1.6174% -1.0966% -0.6405%] (p = 0.00 < 0.05)
                        thrpt:  [+0.6446% +1.1088% +1.6440%]
                        Change within noise threshold.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe
```

**Comment by charliermarsh** [comment]
Merging as a non-behavior-changing refactor that I have high confidence in and don't want to distract others by. If anyone is eager to review I'll happily address follow-up feedback.

**Comment by MichaReiser** [comment]
This new data structure has similarities to our data flow analysis. Do you see way how we could merge the two? 

Out of scope: It would be nice if we only built up these more advanced analyses if a rule consuming the data is enabled.

File: crates/ruff_python_semantic/src/branches.rs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+///     pass
+/// elif x > 1:
+///     pass
+/// else:
+///     pass
+/// ```
+///
+/// Each of the three arms of the `if`-`elif`-`else` would be considered a branch, and would be
+/// assigned their own unique [`BranchId`].
+#[newtype_index]
```
Comment (konstin) [inline crates/ruff_python_semantic/src/branches.rs:19]: i don't like that the macro adds the fields to the struct, it makes it hard to follow and hard to IDE/tool-inspect what data that struct actually is
